pyquil/tests/test_operator_estimation.py

```diff
@@ -599,3 +600,259 @@ def test_ratio_variance_array():
+
+    for idx, res in enumerate(measure_observables(qc,
+                                                  tomo_experiment,
+                                                  n_shots=1000,
+                                                  active_reset=True,
+                                                  readout_symmetrize='exhaustive',
+                                                  calibrate_readout='plus-eig')):
+        raw_e[idx] = res.raw_expectation
+        obs_e[idx] = res.expectation
+        cal_e[idx] = res.calibration_expectation
+
+    assert np.isclose(np.mean(raw_e), 0.849, atol=2e-2)
+    assert np.isclose(np.mean(obs_e), 1.0, atol=2e-2)
+    assert np.isclose(np.mean(cal_e), 0.849, atol=2e-2)
+
+
+def test_exhaustive_symmetrization_1q(forest):
+    qc = get_qc('9q-qvm')
+    qubs = [5]
+    n_shots = 10000
+    p = Program()
+    p00, p11 = 0.90, 0.80
+    p.define_noisy_readout(5, p00, p11)
+    bs_results, d_qub_idx = _exhaustive_symmetrization(qc, qubs, n_shots, p)
+    frac0 = np.count_nonzero(bs_results == 0) / n_shots
+    expected_frac0 = (p00 + p11) / 2
+
+    assert d_qub_idx == {5: 0}
+    assert np.isclose(frac0, expected_frac0, 2e-2)
+
+
+def test_exhaustive_symmetrization_2q(forest):
+    qc = get_qc('9q-qvm')
+    qubs = [5, 7]
+    n_shots = 10000
+    p = Program()
+    p5_00, p5_11 = 0.90, 0.80
+    p7_00, p7_11 = 0.99, 0.77
+    p.define_noisy_readout(5, p5_00, p5_11)
+    p.define_noisy_readout(7, p7_00, p7_11)
+
+    bs_results, d_qub_idx = _exhaustive_symmetrization(qc, qubs, n_shots, p)
+
+    assert d_qub_idx == {5: 0, 7: 1}
+
+    frac5_0 = np.count_nonzero(bs_results[:, d_qub_idx[5]] == 0) / n_shots
+    frac7_0 = np.count_nonzero(bs_results[:, d_qub_idx[7]] == 0) / n_shots
+
+    expected_frac5_0 = (p5_00 + p5_11) / 2
+    expected_frac7_0 = (p7_00 + p7_11) / 2
+
+    assert np.isclose(frac5_0, expected_frac5_0, 2e-2)
+    assert np.isclose(frac7_0, expected_frac7_0, 2e-2)
+
+
+def _random_unitary(n):
+    """
+    :return: array of shape (N, N) representing random unitary matrix drawn from Haar measure
+    """
+    # draw complex matrix from Ginibre ensemble
+    z = np.random.randn(n, n) + 1j * np.random.randn(n, n)
+    # QR decompose this complex matrix
+    q, r = np.linalg.qr(z)
+    # make this decomposition unique
+    d = np.diagonal(r)
+    l = np.diag(d) / np.abs(d)
+    return np.matmul(q, l)
+
+
+def test_process_dfe_bit_flip(forest):
+    qc = get_qc('9q-qvm')
+    # prepare experiment settings
+    expt1 = ExperimentSetting(TensorProductState(plusX(0)), sX(0))
+    expt2 = ExperimentSetting(TensorProductState(plusY(0)), sY(0))
+    expt3 = ExperimentSetting(TensorProductState(plusZ(0)), sZ(0))
+    expt_list = [expt1, expt2, expt3]
+
+    # prepare noisy channel as program
+    prob = 0.3
+    kraus_ops = [np.sqrt(1 - prob) * np.array([[1, 0], [0, 1]]), np.sqrt(prob) * np.array([[0, 1], [1, 0]])]
+    p = Program()
+    p.defgate("DummyGate", _random_unitary(2))
```

Comment:
Can you explain what the random unitary is doing here? DFE as coded only works when computing fidelity to unitaries from the Clifford group. Moreover, it does not scale well with more general unitaries.

Reply:
This is just used as a Dummy Gate to over-ride with Kraus operators. In other words, the random unitary does not actually get used, but is just there to allow for Kraus operations.

Reply:
This is maybe a personal preference, but I don't like the idea of "over-riding" some standard gate with Kraus operations, hence defining some random unitary and over-riding that instead. But since this seems to be objectionable practice, I've gotten rid of it and just replaced it with the identity operation.

Reply:
As discussed, if the error model is meant to be representative of all applications of an operation (i.e., representative of noise in the real machine), overriding should be the correct way to describe the error model.  I don't think it is reasonable to say that it is possible to access perfect X rotations while at the same time having access to imperfect X rotations for other purposes -- unless that is physically motivated somehow. In essence, overriding should be the rule (in this context), not the exception.